### PR TITLE
Wallets refactor

### PR DESCRIPTION
**Pull request by jackstar12**
While working on the node implementation I noticed that all wallet implementations which work over HTTP have a pretty obvious inefficiency: they create a new `httpx.AsyncClient` for every request. This causes a new connection to be made everytime which adds unnecessary overhead (especially with https). 

Instead, the client should be reused. I also added a `cleanup` method to allow proper cleanup on shutdown.

### CODE REVIEW

**Review by talvasconcelos — APPROVED** [review]

File: lnbits/wallets/lnpay.py [hunk 1]
```diff
@@ -69,13 +72,11 @@ async def create_invoice(
...
 
-        async with httpx.AsyncClient() as client:
-            r = await client.post(
-                f"{self.endpoint}/wallet/{self.wallet_key}/invoice",
-                headers=self.auth,
-                json=data,
-                timeout=60,
-            )
+        r = await self.client.post(
+            "/wallet/{self.wallet_key}/invoice",
```
Comment (dni) [inline lnbits/wallets/lnpay.py:76]: this should be a f string

File: lnbits/wallets/lnpay.py [hunk 2]
```diff
@@ -90,13 +91,11 @@ async def create_invoice(
...
     async def pay_invoice(self, bolt11: str, fee_limit_msat: int) -> PaymentResponse:
-        async with httpx.AsyncClient() as client:
-            r = await client.post(
-                f"{self.endpoint}/wallet/{self.wallet_key}/withdraw",
-                headers=self.auth,
-                json={"payment_request": bolt11},
-                timeout=None,
-            )
+        r = await self.client.post(
+            "/wallet/{self.wallet_key}/withdraw",
```
Comment (dni) [inline lnbits/wallets/lnpay.py:95]: +

File: lnbits/wallets/lnpay.py [hunk 3]
```diff
@@ -155,12 +152,9 @@ async def webhook_listener(self):
...
 
         lntx_id = data["data"]["wtx"]["lnTx"]["id"]
-        async with httpx.AsyncClient() as client:
-            r = await client.get(
-                f"{self.endpoint}/lntx/{lntx_id}?fields=settled", headers=self.auth
-            )
-            data = r.json()
-            if data["settled"]:
-                await self.queue.put(lntx_id)
+        r = await self.client.get("/lntx/{lntx_id}?fields=settled")
```
Comment (dni) [inline lnbits/wallets/lnpay.py:155]: +

File: lnbits/wallets/lntips.py [hunk 4]
```diff
@@ -111,11 +108,9 @@ async def pay_invoice(self, bolt11: str, fee_limit_msat: int) -> PaymentResponse
 
     async def get_invoice_status(self, checking_id: str) -> PaymentStatus:
         try:
-            async with httpx.AsyncClient() as client:
-                r = await client.post(
-                    f"{self.endpoint}/api/v1/invoicestatus/{checking_id}",
-                    headers=self.auth,
-                )
+            r = await self.client.post(
+                "/api/v1/invoicestatus/{checking_id}",
```
Comment (dni) [inline lnbits/wallets/lntips.py:112]: +

File: lnbits/wallets/opennode.py [hunk 5]
```diff
@@ -106,9 +106,7 @@ async def pay_invoice(self, bolt11: str, fee_limit_msat: int) -> PaymentResponse
 
     async def get_invoice_status(self, checking_id: str) -> PaymentStatus:
         async with httpx.AsyncClient() as client:
-            r = await client.get(
-                f"{self.endpoint}/v1/charge/{checking_id}", headers=self.auth
-            )
+            r = await client.get("/v1/charge/{checking_id}")
```
Comment (dni) [inline lnbits/wallets/opennode.py:109]: f

File: lnbits/wallets/opennode.py [hunk 6]
```diff
@@ -117,9 +115,7 @@ async def get_invoice_status(self, checking_id: str) -> PaymentStatus:
 
     async def get_payment_status(self, checking_id: str) -> PaymentStatus:
         async with httpx.AsyncClient() as client:
-            r = await client.get(
-                f"{self.endpoint}/v1/withdrawal/{checking_id}", headers=self.auth
-            )
+            r = await client.get("/v1/withdrawal/{checking_id}")
```
Comment (dni) [inline lnbits/wallets/opennode.py:118]: f

File: lnbits/wallets/lnpay.py (hunk 1 above)
Comment (dni) [reply lnbits/wallets/lnpay.py:76]: or am i missing something? seems like all over the code?

**Comment by motorina0** [comment]
Tests should be ran after the below PR has been merged in `main` and `main` rebased on this branch:
https://github.com/lnbits/lnbits/pull/1761/files

**Comment by dni** [comment]
pr is merged

**Comment by motorina0** [comment]
- [spark.py L235](https://github.com/lnbits/lnbits/blob/9b2ecc254f02a7523b5d8f559bf42a2e02d31338/lnbits/wallets/spark.py#L235) still has a `async with httpx.AsyncClient`
- opennode.py has not been changed

File: lnbits/wallets/lnpay.py (hunk 1 above)
Comment (jackstar12) [reply lnbits/wallets/lnpay.py:76]: seems like I overlooked those in my workflow. has to be adjusted

**Comment by jackstar12** [comment]
tests are now actually working and catching the errors

**Comment by jackstar12** [comment]
all wallet functionality is now being tested

File: lnbits/wallets/lnd_grpc_files/router_pb2.py [hunk 7]
```diff
@@ -553,7 +553,6 @@
 
 _ROUTER = DESCRIPTOR.services_by_name["Router"]
 if _descriptor._USE_C_DESCRIPTORS == False:
-
     DESCRIPTOR._options = None
```
Comment (dni) [inline lnbits/wallets/lnd_grpc_files/router_pb2.py:556]: dont touch the lnd grpc files. also some other files only got a blank line removed from some blackking?

File: lnbits/wallets/lndrest.py [hunk 8]
```diff
@@ -172,7 +167,7 @@ async def get_payment_status(self, checking_id: str) -> PaymentStatus:
         except ValueError:
             return PaymentStatus(None)
 
-        url = f"{self.endpoint}/v2/router/track/{checking_id}"
+        url = "/v2/router/track/{checking_id}"
```
Comment (dni) [inline lnbits/wallets/lndrest.py:170]: fstring

File: lnbits/wallets/lndrest.py [hunk 9]
```diff
@@ -183,52 +178,46 @@ async def get_payment_status(self, checking_id: str) -> PaymentStatus:
...
+                        return PaymentStatus(None)
+                except:
+                    continue
 
         return PaymentStatus(None)
 
     async def paid_invoices_stream(self) -> AsyncGenerator[str, None]:
         while True:
             try:
                 url = self.endpoint + "/v1/invoices/subscribe"
```
Comment (dni) [inline lnbits/wallets/lndrest.py:209]: still self.endpoint here?

File: lnbits/wallets/opennode.py [hunk 10]
```diff
@@ -61,17 +62,15 @@ async def create_invoice(
...
-                    # "callback_url": url_for("/webhook_listener", _external=True),
-                },
-                timeout=40,
-            )
+        r = await self.client.post(
+            "/v1/charges",
+            json={
+                "amount": amount,
+                "description": memo or "",
+                # "callback_url": url_for("/webhook_listener", _external=True),
```
Comment (motorina0) [inline lnbits/wallets/opennode.py:70]: dead code, should be removed

**Review by callebtc — APPROVED** [review]
LGTM

Did you make any other logic changes except for the constants in `lnbits.py:L99`?

**Pull request merged**